### PR TITLE
Add Users API

### LINKED ISSUE #4: Add User endpoints.
#  Add user endpoints

- `POST /signup` Signup user
- `POST /login` login user
- `POST /verify/{code}` verify the user's email
- `GET /user/me`: Get me information 
- `GET /user/{user_id)`: Get user information 
- `PUT /user/me`: Update user's info 
- `Delete  /user/me`: Delete a user

### PR DESCRIPTION
**Pull request by 7adidaz**
# Description

Fixes #4 

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: api-library/api.yaml [hunk 1]
```diff
@@ -186,6 +188,183 @@ paths:
...
+            application/json: 
+              schema:
+                $ref: '#/components/schemas/User'
+        
+        "422": 
+          description: validation error
+          content:
+            application/json: 
+              schema:
+                $ref: '#/components/schemas/User'
```
[the remark this answers is not in the document]
Comment (7adidaz) [reply api-library/api.yaml:247]: ACK

File: api-library/api.yaml [hunk 2]
```diff
@@ -370,8 +549,60 @@ components:
...
+          type: string
+        online_status:
+          type: string
+          enum:
+            - online
+            - away
+            - offline
+        created_at:
+          type: string
+          format: date
```
Comment (mohammedzaher) [inline api-library/api.yaml:605]: I don't get why you defined UserForUsers schema

File: api-library/api.yaml [hunk 3]
```diff
@@ -186,6 +188,183 @@ paths:
...
+      responses:
+        "200": 
+          $ref: '#/components/responses/ok'
+        
+        "422": 
+          description: user's loggin credentioals didn't pass the validation step 
+          content:
+            application/json: 
+              schema:
+                $ref: '#/components/schemas/UserAuth'
```
Comment (mohammedzaher) [inline api-library/api.yaml:360]: I think this could be generalized in the responses section.

File: api-library/api.yaml [hunk 4]
```diff
@@ -186,6 +188,183 @@ paths:
...
```
Comment (mohammedzaher) [inline api-library/api.yaml:331]: I think this could also be generalized in the responses section.

**Review by mohammedzaher — CHANGES_REQUESTED** [review]

**Review by moaazelsayed1 — CHANGES_REQUESTED** [review]

File: api-library/api.yaml (hunk 4 above)
Comment (7adidaz) [reply api-library/api.yaml:331]: should it? it's only being mentioned twice.

File: api-library/api.yaml [hunk 5]
```diff
@@ -186,6 +188,183 @@ paths:
...
+                
+        
+        "401": 
+          $ref: '#/components/responses/Unauthorized'
+          
+        "404": 
+          $ref: '#/components/responses/NotFound'
   /threads:
```
[the remark this answers is not in the document]
Comment (7adidaz) [reply api-library/api.yaml:368]: ack

File: api-library/api.yaml (hunk 2 above)
Comment (7adidaz) [reply api-library/api.yaml:605]: ack.

File: api-library/api.yaml [hunk 6]
```diff
@@ -186,6 +188,183 @@ paths:
...
+          description: user's loggin credentioals didn't pass the validation step 
+          content:
+            application/json: 
+              schema:
+                $ref: '#/components/schemas/UserAuth'
+                
+        
+        "401": 
+          $ref: '#/components/responses/Unauthorized'
+
```
[the remark this answers is not in the document]
Comment (7adidaz) [reply api-library/api.yaml:365]: ack

**Comment by 7adidaz** [comment]
@moaazelsayed1  don't forget to approve to merge.

**Review by moaazelsayed1 — APPROVED** [review]

**Review by mohammedzaher — COMMENTED** [review]
Ok let's finish this review Good Work

**Pull request merged**